### PR TITLE
Support skip_unused_shards in Analyzer

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit 7c0b6b4d900be466d82dbfdb9bc9d9a14ea3baa9 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53282/7c0b6b4d900be466d82dbfdb9bc9d9a14ea3baa9/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Interpreters/evaluateConstantExpression.cpp [hunk 1]
```diff
@@ -368,6 +376,339 @@ namespace
...
+
+std::optional<ConstantVariants> evaluateExpressionOverConstantCondition(
+    const ActionsDAG::Node * predicate,
+    const ActionsDAG::NodeRawConstPtrs & expr,
+    const ContextPtr & context,
+    size_t max_elements)
+{
+    auto inverted_dag = KeyCondition::cloneASTWithInversionPushDown({predicate}, context);
+    // std::cerr << "--- inverted dag\n";
+    // std::cerr << inverted_dag->dumpDAG() << std::endl;
```
Comment (novikd) [inline src/Interpreters/evaluateConstantExpression.cpp:684]:
```suggestion
```

File: src/Interpreters/evaluateConstantExpression.cpp [hunk 2]
```diff
@@ -368,6 +376,339 @@ namespace
...
+    auto matches = matchTrees(expr, *inverted_dag, false);
+
+    // for (const auto & [node, match] : matches)
+    // {
+    //     LOG_TRACE(&Poco::Logger::get("eeocc"), "Match {} {} -> {} {} (with monotonicity : {})",
+    //         static_cast<const void *>(node), node->result_name,
+    //         static_cast<const void *>(match.node), (match.node ? match.node->result_name : ""), match.monotonicity != std::nullopt);
+    // }
```
Comment (novikd) [inline src/Interpreters/evaluateConstantExpression.cpp:692]:
```suggestion
```

File: src/Interpreters/evaluateConstantExpression.cpp [hunk 3]
```diff
@@ -368,6 +376,339 @@ namespace
...
+
+    auto predicates = analyze(inverted_dag->getOutputs().at(0), matches, context, max_elements);
+
+    if (!predicates)
+        return {};
+
+    // std::cerr << "--- num predicates " << predicates->size() << std::endl;
```
Comment (novikd) [inline src/Interpreters/evaluateConstantExpression.cpp:699]:
```suggestion
```

File: src/Storages/StorageDistributed.cpp [hunk 4]
```diff
@@ -512,6 +516,100 @@ QueryProcessingStage::Enum StorageDistributed::getQueryProcessingStage(
...
+        (settings.allow_nondeterministic_optimize_skip_unused_shards || sharding_key_is_deterministic);
+
+    QueryProcessingStage::Enum default_stage = QueryProcessingStage::WithMergeableStateAfterAggregation;
+    if (settings.distributed_push_down_limit)
+        default_stage = QueryProcessingStage::WithMergeableStateAfterAggregationAndLimit;
+
+    const auto & query_node = query_info.query_tree->as<const QueryNode &>();
+
+    // std::cerr << query_node.dumpTree() << std::endl;
+    // std::cerr << query_info.table_expression->dumpTree() << std::endl;
```
Comment (novikd) [inline src/Storages/StorageDistributed.cpp:534]:
```suggestion
```

File: src/Storages/StorageDistributed.cpp [hunk 5]
```diff
@@ -512,6 +516,100 @@ QueryProcessingStage::Enum StorageDistributed::getQueryProcessingStage(
...
+    if (query_node.isDistinct())
+    {
+        if (!optimize_sharding_key_aggregation || !expr_contains_sharding_key(query_node.getProjection()))
+            return {};
+    }
+
+    // GROUP BY
+    bool has_aggregates = query_info.has_aggregates;
+    if (query_info.syntax_analyzer_result)
+        has_aggregates = !query_info.syntax_analyzer_result->aggregates.empty();
```
Comment (novikd) [inline src/Storages/StorageDistributed.cpp:585]: strange to see in new infrastructure

File: src/Storages/StorageDistributed.cpp [hunk 6]
```diff
@@ -1360,6 +1465,76 @@ IColumn::Selector StorageDistributed::createSelector(const ClusterPtr cluster, c
...
+    std::set<int> shards;
+
+    for (const auto & variant : *variants)
+    {
+        const auto selector = createSelector(cluster, variant.at(0));
+        shards.insert(selector.begin(), selector.end());
+    }
+
+    // for (int i : shards)
+    //     std::cerr << ".. shard " << i << std::endl;
```
Comment (novikd) [inline src/Storages/StorageDistributed.cpp:1533]:
```suggestion
```

File: src/Storages/StorageDistributed.cpp [hunk 7]
```diff
@@ -1360,6 +1465,76 @@ IColumn::Selector StorageDistributed::createSelector(const ClusterPtr cluster, c
...
+
+    const auto & sharding_key_dag = sharding_key_expr->getActionsDAG();
+    const auto * expr_node = sharding_key_dag.tryFindInOutputs(sharding_key_column_name);
+    if (!expr_node)
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR, "Cannot find sharding key column {} in expression {}",
+            sharding_key_column_name, sharding_key_dag.dumpDAG());
+
+    // std::cerr << "--- expr\n";
+    // std::cerr << sharding_key_dag.dumpDAG() << std::endl;
```
Comment (novikd) [inline src/Storages/StorageDistributed.cpp:1514]:
```suggestion
```

**Review by novikd — COMMENTED** [review]
In general LGTM

**Comment by KochetovNicolai** [comment]
test_s3_engine_heavy_write_check_mem is flaky and not fixed yet

[test_chroot is flaky](https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICcldGVzdF9jaHJvb3QlJwpPUkRFUiBCWSBjaGVja19zdGFydF90aW1lIGRlc2MsIGNoZWNrX25hbWUsIHRlc3RfbmFtZQ==)

**Pull request merged**